miner/algo_greedy_buckets.go

```diff
@@ -0,0 +1,195 @@
+	transactions []*types.TxWithMinerFee,
+	orders *types.TransactionsByPriceAndNonce,
+	retryMap map[*types.TxWithMinerFee]int, retryLimit int,
+) ([]types.SimulatedBundle, []types.UsedSBundle) {
+	var (
+		usedBundles                []types.SimulatedBundle
+		usedSbundles               []types.UsedSBundle
+		CheckRetryOrderAndReinsert = func(order *types.TxWithMinerFee, orders *types.TransactionsByPriceAndNonce, retryMap map[*types.TxWithMinerFee]int, retryLimit int) {
+			var isRetryable bool
+			if retryCount, exists := retryMap[order]; exists {
+				if retryCount != retryLimit {
+					isRetryable = true
+					retryMap[order] = retryCount + 1
+				}
+			} else {
+				retryMap[order] = 0
+				isRetryable = true
+			}
+
+			if isRetryable {
+				orders.Push(order)
+			}
+		}
+	)
+
+	for _, order := range transactions {
+		if tx := order.Tx(); tx != nil {
+			receipt, skip, err := envDiff.commitTx(tx, b.chainData)
+			if err != nil {
+				log.Trace("could not apply tx", "hash", tx.Hash(), "err", err)
+				CheckRetryOrderAndReinsert(order, orders, retryMap, retryLimit)
+				continue
+			}
+
+			if skip == shiftTx {
+				orders.ShiftAndPushByAccountForTx(tx)
+			}
+
+			effGapPrice, err := tx.EffectiveGasTip(envDiff.baseEnvironment.header.BaseFee)
+			if err == nil {
+				log.Trace("Included tx", "EGP", effGapPrice.String(), "gasUsed", receipt.GasUsed)
+			}
+		} else if bundle := order.Bundle(); bundle != nil {
+			err := envDiff.commitBundle(bundle, b.chainData, b.interrupt, true)
+			if err != nil {
+				log.Trace("Could not apply bundle", "bundle", bundle.OriginalBundle.Hash, "err", err)
+				CheckRetryOrderAndReinsert(order, orders, retryMap, retryLimit)
+				continue
+			}
+
+			log.Trace("Included bundle", "bundleEGP", bundle.MevGasPrice.String(),
+				"gasUsed", bundle.TotalGasUsed, "ethToCoinbase", ethIntToFloat(bundle.TotalEth))
+			usedBundles = append(usedBundles, *bundle)
+		} else if sbundle := order.SBundle(); sbundle != nil {
+			usedEntry := types.UsedSBundle{
+				Bundle: sbundle.Bundle,
+			}
+			err := envDiff.commitSBundle(sbundle, b.chainData, b.interrupt, b.builderKey, true)
+			if err != nil {
+				log.Trace("Could not apply sbundle", "bundle", sbundle.Bundle.Hash(), "err", err)
+				// TODO: handle retry
+				usedEntry.Success = false
+				usedSbundles = append(usedSbundles, usedEntry)
+				CheckRetryOrderAndReinsert(order, orders, retryMap, retryLimit)
+				continue
+			}
+
+			log.Trace("Included sbundle", "bundleEGP", sbundle.MevGasPrice.String(), "ethToCoinbase", ethIntToFloat(sbundle.Profit))
+			usedEntry.Success = true
+			usedSbundles = append(usedSbundles, usedEntry)
+		}
+	}
+	return usedBundles, usedSbundles
+}
+
+func (b *greedyBucketsBuilder) mergeOrdersIntoEnvDiff(
+	envDiff *environmentDiff, orders *types.TransactionsByPriceAndNonce) ([]types.SimulatedBundle, []types.UsedSBundle) {
+	if orders.Peek() == nil {
+		return nil, nil
+	}
+
+	const (
+		retryLimit = 2
```

Comment:
This seems huge, since most orders won't "get fixed", and most orders will fail.
I'd probably suggest only retrying once, and only if the error is "bundle underpays"

Reply:
updated retry limit